Persist prompt dismissal in user preferences

Jinksi opened the pull request:
Resolves #10329

> [!IMPORTANT]
> Based on branch from PR #10326 which introduces the prompt. Merge it first. 


> [!IMPORTANT]
> Awaiting other user interaction sources that will be used to trigger dismissal:
>	* #10327
>	* #10328

#### Changes proposed in this Pull Request

* Stores the prompt dismissal state in WooCommerce user preferences meta, specifically the user meta key `woocommerce_admin_wc_payments_wporg_review_2025_prompt_dismissed`.
* Users will not see the prompt again if they've:
	* Clicked `Yes`
	* Clicked `No`
	* Clicked `X` or the body of the snackbar.
* This dismissal state is stored separately for each WP admin user, meaning other users in the same store's wp-admin can still be prompted for a review.
* The UNIX timestamp of the dismiss request is stored instead of just a boolean value, in case we want to allow it to expire at some point.



#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Ensure your store is eligible to see the prompt as per the criteria:
	* Minimum 5 payouts.
	* Minimum $1,000 TPV
	* Not rejected or suspended.
* Enable the client dev feature flag `_wcpay_feature_prompt_merchant_for_review`
	```php
	update_option( '_wcpay_feature_prompt_merchant_for_review', '1' );
	```
* Dismiss the prompt using `X` or clicking the body, or interact with the prompt using `Yes` or `No`. Ensure the prompt is dismissed and **does not re-appear** on navigation or page reload.
* After dismissal, ensure that the prompt **does appear** when logged in as a separate user.
* For testing purposes, you can reset the dismissal state in user preferences by running:
	```php
	delete_user_meta( get_current_user_id(), 'woocommerce_admin_wc_payments_wporg_review_2025_prompt_dismissed');
	```

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : **QA Testing Not Applicable** since eligibility is difficult to prepare, manual internal testing is suitable.
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable. **Not a critical flow**
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable. **What's new in 9.1.0 post: paJDYF-gXj-p2**

Jinksi committed:
Add support for WordPress.org Review 2025 campaign flag

Jinksi committed:
Add comment to clarify WordPress.org Review 2025 campaign flag behavior

Jinksi committed:
Add merchant feedback prompt feature flag

Jinksi committed:
Add merchant feedback prompt feature flag to wcpaySettings

Jinksi committed:
Add initial merchant feedback prompt component and integration

- Create merchant feedback prompt component
- Show merchant feedback prompt on overview page with feature flag

Jinksi committed:
Add feedback prompt with dismissal functionality, persisted in local component state

Jinksi committed:
Show prompt on many Payments screens

Jinksi committed:
Remove test notice

Jinksi committed:
Merge branch 'develop' into add/10326-add-prompt-for-review

Jinksi committed:
Refactor: Improve merchant feedback prompt portal documentation and simplify portal rendering

Jinksi committed:
Rename SnackbarPortal to WCFooterPortal for clarity

Jinksi committed:
Add comment explaining manual explicit dismiss button

Jinksi committed:
Record tracks events on interactions

Jinksi committed:
Dismiss prompt on element click, consistent with other snackbars

Jinksi committed:
Add integration testing of component

Jinksi committed:
Simplify by moving dev feature flag check to MerchantFeedbackPrompt component

Jinksi committed:
Rename merchant feedback prompt feature flag to clarify development purpose

Jinksi committed:
Add code comments explaining the campaign flag

Jinksi committed:
Add account campaign eligibility check to merchant feedback prompt

Jinksi committed:
Add placeholder changelog entry

Jinksi committed:
Show prompt on payment details screen

Jinksi committed:
Refactor merchant feedback prompt to use wrapper component

Jinksi committed:
Add comments explaining temporary snackbar implementation details

Jinksi committed:
Persist prompt dismiss in user preferences

Jinksi committed:
Clarify the `dismissPrompt` callback function

Jinksi committed:
Merge branch 'add/10326-add-prompt-for-review' into update/10329-persist-user-prompt-dismiss

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10444` or branch name `update/10329-persist-user-prompt-dismiss` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=update/10329-persist-user-prompt-dismiss) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: eca1a56464ab3d639b4a2dc998395b22005c04c8
- Build time: 2025-03-05 08:11:36 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi committed:
Update tests with mock useUserPreferences

jessy-p committed:
Merge branch 'develop' into add/10326-add-prompt-for-review

Jinksi committed:
Merge branch 'develop' into add/10326-add-prompt-for-review

Jinksi committed:
Merge branch 'add/10326-add-prompt-for-review' into update/10329-persist-user-prompt-dismiss

Jinksi committed:
Remove unused useState import

Jinksi committed:
Encapsulate prompt dismissal state management to react hook

Jinksi committed:
Update merchant feedback prompt to use timestamp for dismissal

Jinksi committed:
Add placeholder changelog entry

Jinksi commented:
Bringing this back to **In progress**, while we await other user interaction sources that will be used to trigger dismissal:
* #10327
* #10328

Jinksi committed:
Add more details explaining the Snackbar and portal workarounds

Jinksi committed:
Merge branch 'develop' into add/10326-add-prompt-for-review

Jinksi committed:
Fix missing `>` after merge conflict resolution

Jinksi committed:
Change component name to MaybeShowMerchantFeedbackPrompt to clarify behaviour

Jinksi committed:
Simplify merchant feedback prompt view event tracking

Remove unnecessary ref tracking for view event recording, ensuring the event is tracked directly when the prompt should be shown

Jinksi committed:
Merge branch 'add/10326-add-prompt-for-review' into update/10329-persist-user-prompt-dismiss

Jinksi committed:
Remove unused `useRef` import

Jinksi commented on client/merchant-feedback-prompt/hooks.ts:
```diff
@@ -0,0 +1,48 @@
+		...userPrefs
+	} = useUserPreferences() as UserPreferences;
+
+	const isAccountEligible =
+		wcpaySettings?.featureFlags?.isMerchantFeedbackPromptDevFlagEnabled &&
+		wcpaySettings?.accountStatus?.campaigns?.wporgReview2025;
```
As @haszari  mentioned in https://github.com/Automattic/woocommerce-payments/pull/10370/files#r1976770261, consider using early return logic to make this easier to read.


```js
if ( the store is a prod store and not opted in to testing our code before we're ready to release ) { return and render nothing }:

if ( the store is not within the campaign criteria ) { return and render nothing }:

if ( the merchant has dismissed or responded ) { return and render nothing }:

// (else)

// ask for feedback
```

Jinksi committed:
Simplify and clarify temporary workaround code comments

Jinksi committed:
Clarify feature flag code comment

Jinksi committed:
Update comment for merchant feedback prompt temporary style

Jinksi committed:
Link to issue in dev feature flag code comment

Jinksi committed:
Merge branch 'add/10326-add-prompt-for-review' into update/10329-persist-user-prompt-dismiss

Jinksi committed:
Merge branch 'develop' into update/10329-persist-user-prompt-dismiss

jessy-p committed:
Merge branch 'develop' into update/10329-persist-user-prompt-dismiss

jessy-p reviewed (approved):
Tested that the modal is dismissed on clicking 'Yes', 'No' and  'x' ✅ 
Tested with different users ✅ 

Note: If am deleting the usermeta from the DB, I need to take care not to be on a payments page (otherwise it persists the metadata again)
